packages/block-editor/src/hooks/style.js

```diff
@@ -133,14 +141,35 @@ export function addSaveProps( props, blockType, attributes ) {
 		[ COLOR_SUPPORT_KEY ]: getBlockSupport( blockType, COLOR_SUPPORT_KEY ),
 	} );
 
+	let typographyFeaturesToRemove = [];
+	if (
+		getBlockSupport(
+			blockType,
+			'__experimentalSkipTypographySerialization'
+		)
+	) {
+		// Font size is handled separately.
+		typographyFeaturesToRemove = without(
+			TYPOGRAPHY_SUPPORT_KEYS,
+			FONT_SIZE_SUPPORT_KEY
+		);
+	}
+
 	if (
 		getBlockSupport( blockType, '__experimentalSkipFontSizeSerialization' )
 	) {
-		filteredStyle = omit( filteredStyle, [
-			[ 'typography', FONT_SIZE_SUPPORT_KEY ],
-		] );
+		typographyFeaturesToRemove = [
+			...typographyFeaturesToRemove,
+			FONT_SIZE_SUPPORT_KEY,
+		];
 	}
 
+	const { typography, ...otherStyle } = filteredStyle;
```

Comment:
We are doing this in two different ways: 

- using `omitKeysNotToSerialize` for some properties (and the reader of this function is abstracted from the details)
- directly filtering the fontSize and rest of typography in the function

I'm fine with using any approach as far as it's the same for all properties.

---

What would you think of consolidating into something like (pseudo-code):

```js
let filteredStyle = omitKeysNotToSerialize( style, {
    border: getBlockSupport( blockType, BORDER_SUPPORT_KEY ).__experimentalSkipSerialization,
    [ COLOR_SUPPORT_KEY ]: getBlockSupport( blockType, COLOR_SUPPORT_KEY ).__experimentalSkipSerialization,
    fontSize: getBlockSupport( blockType, '__experimentalSkipFontSizeSerialization' ),
    textDecoration: getBlockSupport( blockType, '__experimentalSkipTypographySerialization' ),
    textTransform: getBlockSupport( blockType, '__experimentalSkipTypographySerialization' )
} );
```

so the data passed to `omitKeysNotToSerialize` is already a boolean.

Reply:
> We are doing this in two different ways:
> 
> * using `omitKeysNotToSerialize` for some properties (and the reader of this function is abstracted from the details)
> * directly filtering the fontSize and rest of typography in the function
> 
> I'm fine with using any approach as far as it's the same for all properties.

Yeah, it's a bit messy right now 😬 

> What would you think of consolidating into something like (pseudo-code):
> 
> ```js
> let filteredStyle = omitKeysNotToSerialize( style, {
>     border: getBlockSupport( blockType, BORDER_SUPPORT_KEY ).__experimentalSkipSerialization,
>     [ COLOR_SUPPORT_KEY ]: getBlockSupport( blockType, COLOR_SUPPORT_KEY ).__experimentalSkipSerialization,
>     fontSize: getBlockSupport( blockType, '__experimentalSkipFontSizeSerialization' ),
>     textDecoration: getBlockSupport( blockType, '__experimentalSkipTypographySerialization' ),
>     textTransform: getBlockSupport( blockType, '__experimentalSkipTypographySerialization' )
> } );
> ```
> 
> so the data passed to `omitKeysNotToSerialize` is already a boolean.

I like the shape, but I'm once more worried that it'd be very easy to miss for someone who adds a new typographic feature (e.g. letter spacing, see #31118). We'd be effectively duplicating information found in `TYPOGRAPHY_SUPPORT_KEYS` -- which is otherwise used a lot in the context of the block-supports mechanism, so it would come as somewhat of a surprise.

I'd rather keep using `TYPOGRAPHY_SUPPORT_KEYS` (to centralize the source of truth -- with the grain of salt that font size is exempt here). I had originally tried something along the lines of your suggestion, but that doesn't really work with the `TYPOGRAPHY_SUPPORT_KEYS` const 😬 

I'll see if I can find an alternative 🤔

Reply:
Ahh, I had missed one thing when replying 😬  -- the keys we need here (e.g. `textTransform`) aren't identical to the values in `TYPOGRAPHY_SUPPORT_KEYS` (where some of them have the `__experimental` prefix, e.g. `__experimentalTextTransform`). That's also the reason for [this issue](https://github.com/WordPress/gutenberg/pull/30880#issuecomment-836668519) that you found.

So we can't leave things exactly as they are but will at least to fix this. And maybe your suggestion is then the most straight-forward fix 😅

Reply:
Hmm no, that doesn't work either, unfortunately 😕  The problem is that the `typography` style attributes are nested, so a line like

```js
textDecoration: getBlockSupport( blockType, '__experimentalSkipTypographySerialization' ),
```

doesn't work, since we don't want to remove a top-level attribute, but the `textDecoration` attribute of the `typography` object.

Unfortunately, the obvious alternative doesn't work either:

```js
'typography.textTransform': getBlockSupport( blockType, '__experimentalSkipTypographySerialization' ),
```

The reason is that unlike `omit`, `omitBy` doesn't support object "path" notation. (We probably extend `omitKeysNotToSerialize` to interpret keys that contain a period `.` as paths, but that makes things a bit too magical, IMO.)

I'll try something closer to my original approach...

Reply:
Okay, this required a rather far-reaching refactor of the relevant code, but I think this should be working now: bb5b6f2